### PR TITLE
Assigning numeric ID for problems

### LINKED ISSUE #330: Support assigning numeric id for problems via web interfaces.

### PR DESCRIPTION
**Pull request by moesoha**
Show a checkbox "assign numeric ID" when creating problems, default value is True.

Close #330

### CODE REVIEW

File: vj4/model/domain.py [hunk 1]
```diff
@@ -106,6 +106,26 @@ def get_pending(**kwargs):
...
+
+  Returns:
+    Integer value after increment.
+  """
+  coll = db.coll('domain')
+  doc = await coll.find_one_and_update(filter={'_id': domain_id},
+                                       update={'$inc': {'pid_counter': 1}},
+                                       upsert=True,
+                                       return_document=ReturnDocument.AFTER)
+  if doc['pid_counter'] == 1:
```
Comment (twd2) [inline vj4/model/domain.py]: Bug if somehow 2 or more requests reach this point concurrently.

Comment (iceb0y) [reply vj4/model/domain.py]: You can use `$setOnInsert` to set the default value. See https://docs.mongodb.com/manual/reference/operator/update/setOnInsert/ .

However, a unique index is needed for atomicity. Also see discussion at https://grokbase.com/t/gg/mongodb-user/14aeaptjb7/findandupdate-atomicity-without-a-unique-index-im-stuck .

Comment (iceb0y) [reply vj4/model/domain.py]: Actually `upsert` and the unique index is not needed because domain doc must exist and `$setOnInsert` should not work.

Comment (iceb0y) [reply vj4/model/domain.py]: I think the solution here should work but it's crowded
https://stackoverflow.com/questions/29746158/mongodb-counter-with-initial-value-not-zero

Other alternatives:
* add 1000 to the return value
* start from 1
* update the database offline

Comment (moesoha) [reply vj4/model/domain.py]: Use "set if not exists" finally

File: vj4/model/domain.py [hunk 2]
```diff
@@ -106,6 +106,22 @@ def get_pending(**kwargs):
...
+@argmethod.wrap
+async def inc_pid_counter(domain_id):
+  """Increments the problem ID counter.
+
+  Returns:
+    Integer value after increment.
+  """
+  coll = db.coll('domain')
+  await coll.update_one(filter={'_id': domain_id, 'pid_counter': {'$exists': False}},
+                        update={'$set': {'pid_counter': 999}})
```
Comment (iceb0y) [inline vj4/model/domain.py]: 1000

File: vj4/model/domain.py [hunk 3]
```diff
@@ -106,6 +106,22 @@ def get_pending(**kwargs):
...
+  doc = await coll.find_one_and_update(filter={'_id': domain_id},
+                                       update={'$inc': {'pid_counter': 1}},
```
Comment (iceb0y) [inline vj4/model/domain.py]: remove `return_document=ReturnDocument.AFTER` as discussed

File: vj4/ui/templates/problem_edit.html [hunk 4]
```diff
@@ -13,9 +13,17 @@
               </label>
             </div>
             {% if page_name == 'problem_create' %}
-            <div class="medium-4 columns">
+            <div class="medium-2 columns">
               <label>
                 {{ _('Settings') }}
+                <br>
+                <label class="checkbox">
+                  <input type="checkbox" name="numeric_pid" value="on" checked>{{ _('Numeric PID') }}
```
Comment (iceb0y) [inline vj4/ui/templates/problem_edit.html]: Use/Allocate numeric PID?

Comment (moesoha) [reply vj4/ui/templates/problem_edit.html]: Too long

**Pull request merged**